src/main/java/org/mule/extensions/java/internal/metadata/ExecutableElementTypeResolver.java

```diff
@@ -0,0 +1,161 @@
+import static java.util.Collections.emptySet;
+import static java.util.stream.Collectors.toList;
+import static org.mule.runtime.api.metadata.MetadataKeyBuilder.newKey;
+import static org.mule.runtime.api.metadata.resolving.FailureCode.INVALID_METADATA_KEY;
+import static org.mule.runtime.core.api.util.StringUtils.isBlank;
+import org.mule.extensions.java.internal.JavaModule;
+import org.mule.extensions.java.internal.parameters.ExecutableIdentifier;
+import org.mule.extensions.java.internal.parameters.ExecutableIdentifierFactory;
+import org.mule.metadata.api.builder.ObjectTypeBuilder;
+import org.mule.metadata.api.model.MetadataType;
+import org.mule.runtime.api.connection.ConnectionException;
+import org.mule.runtime.api.metadata.MetadataContext;
+import org.mule.runtime.api.metadata.MetadataKey;
+import org.mule.runtime.api.metadata.MetadataKeyBuilder;
+import org.mule.runtime.api.metadata.MetadataResolvingException;
+import org.mule.runtime.api.metadata.resolving.InputTypeResolver;
+import org.mule.runtime.api.metadata.resolving.OutputTypeResolver;
+import org.mule.runtime.api.metadata.resolving.TypeKeysResolver;
+import org.mule.runtime.core.api.util.ClassUtils;
+
+import java.lang.reflect.Executable;
+import java.lang.reflect.Method;
+import java.lang.reflect.Type;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Base {@link InputTypeResolver}, {@link OutputTypeResolver} and {@link TypeKeysResolver}
+ * for any {@link Executable} element.
+ *
+ * @since 1.0
+ */
+abstract class ExecutableElementTypeResolver implements OutputTypeResolver<ExecutableIdentifier>,
+    InputTypeResolver<ExecutableIdentifier>,
+    TypeKeysResolver {
+
+  @Override
+  public String getCategoryName() {
+    return "MethodTypes";
+  }
+
+  @Override
+  public String getResolverName() {
+    return "MethodTypeResolver";
+  }
+
+  protected abstract List<Executable> getExecutableElements(Class<?> targetClass);
+
+  @Override
+  public MetadataType getInputMetadata(MetadataContext context, ExecutableIdentifier key)
+      throws MetadataResolvingException, ConnectionException {
+
+    Executable element = findElement(key);
+
+    if (element.getParameters().length == 0) {
+      return context.getTypeBuilder().nullType().build();
+    }
+
+    ObjectTypeBuilder inputParameters = context.getTypeBuilder().objectType().id(key.getElementId() + "_INPUT");
+    stream(element.getParameters())
+        .forEach(param -> inputParameters.addField()
+            .key(param.getName())
+            .value(context.getTypeLoader().load(param.getType())));
+
+    return inputParameters.build();
+  }
+
+  @Override
+  public MetadataType getOutputType(MetadataContext context, ExecutableIdentifier key)
+      throws MetadataResolvingException, ConnectionException {
+
+    Executable element = findElement(key);
+    Type output = element instanceof Method
+        ? ((Method) element).getGenericReturnType()
+        : loadClass(key.getClazz());
```

Comment:
type TCCL

Reply:
forgot about this?

Reply:
loadClass does `ClassUtils.loadClass(className, Thread.currentThread().getContextClassLoader())`